crypto/context.c

```diff
@@ -15,6 +15,19 @@ struct openssl_ctx_onfree_list_st {
     struct openssl_ctx_onfree_list_st *next;
 };
 
+extern const OPENSSL_CTX_METHOD default_method_store_method;
+extern const OPENSSL_CTX_METHOD provider_store_method;
+extern const OPENSSL_CTX_METHOD property_defns_method;
+extern const OPENSSL_CTX_METHOD property_string_data_method;
+
+/* Must be in the same order as the indexes in cryptlib.h */
+static const OPENSSL_CTX_METHOD *methods[] = {
+    &default_method_store_method,
+    &provider_store_method,
+    &property_defns_method,
+    &property_string_data_method
+};
```

Comment:
Please no.  This is not a good idea, at all.  If nothing else, then that comment should indicate that this is a maintenance problem waiting to happen.  There's really no reason why code that initiate an index shouldn't know their corresponding `OPENSSL_CTX_METHOD`.

Reply:
An alternative I considered was to pass the method every time we call openssl_ctx_get_data. I thought this approach was nicer but could go that way if you prefer.

Reply:
The idea was that each sub-system using this would use a pattern like this:

``` C
static int init_my_store(OPENSSL_CTX *libctx)
{
    return openssl_ctx_init_index(libctx, MY_INDEX, &my_method);
}

static blah *get_my_store(OPENSSL_CTX *libctx)
{
    if (!openssl_ctx_run_once(libctx, MY_INDEX, init_my_store))
       return NULL;

    /* At this point, the index is correct and its method initiated */
    return openssl_ctx_get_data(libctx, MY_INDEX);
}
```

Anything else would be an invalid implementation.

Reply:
Yes, I know that was the idea. And that's what I had originally. But then I ended up with loads of openssl_ctx_run_once calls and init functions that *only* initialised the index. There was a lot of extra code that essentially did very little. If I got rid of the need for that everything suddenly became a lot simpler.

Reply:
... and that leads to a managment problem.  I find that *more* displeasing